File: src/myplanet_js/myplanet.js
const channelAliases = {
    r: ["r", "R", "red", "rot"],
    g: ["g", "G", "green", "gruen", "grün"],
    b: ["b", "B", "blue", "blau"],
    a: ["a", "A", "alpha", "alpha"],
    elevation: ["h", "H", "elevation", "hoehe", "höhe"],
    country: ["c", "C", "country", "land"],
    pop: ["p", "P", "pop", "bev"],
    nox: ["n", "N", "nox", "NOx"],
    night: ["l", "L", "night", "nacht", "light", "nightlight", "licht"],
};
// @ts-ignore
myplanet = createMyplanetLib();
class MercatorWorker {
    static EarthRadius = 6378137;
    static metersPerPixel(latitude, zoomLevel) {
        return Math.cos(latitude * Math.PI / 180) * 2 * Math.PI * this.EarthRadius / this.getMapSize(zoomLevel);
    }
    static getMapSize(zoomLevel) {
        return Math.ceil(256 * Math.pow(2, zoomLevel));
    }
    static clip(n, minValue, maxValue) {
        return Math.min(Math.max(n, minValue), maxValue);
    }
    static worldPixelToPosition(px, py, zoomLevel) {
        var mapSize = this.getMapSize(zoomLevel);
        var x = (this.clip(px, 0, mapSize - 1) / mapSize) - 0.5;
        var y = 0.5 - (this.clip(py, 0, mapSize - 1) / mapSize);
        return [
            360 * x,
            90 - 360 * Math.atan(Math.exp(-y * 2 * Math.PI)) / Math.PI //Latitude
        ];
    }
}
function createMyplanetLib() {
    const asyncWaitFor = async (signal) => {
        let answer = waitFor(signal);
        while (!answer) {
            await new Promise(r => setTimeout(r, 200));
            answer = waitFor(signal);
        }
        return answer.payload;
    };
    let globalChannelkeys = [];
    class Messenger {
        async imageWaitFor(signal) {
            return asyncWaitFor(signal);
        }
        async runCommandAsync(command, value) {
            const signal = 'image' + command + Date.now();
            show({ 'command': command, 'value': value, 'readysignal': signal });
            return await this.imageWaitFor(signal);
        }
    }
    class LocatedMessenger extends Messenger {
        zoomLevel;
        worldCoords;
        W;
        H;
        getPosition(px, py) {
            return MercatorWorker.worldPixelToPosition(this.worldCoords[0] + px, this.worldCoords[1] + py, this.zoomLevel);
        }
        metersPerPixel(px, py) {
            return MercatorWorker.metersPerPixel(this.getPosition(px, py)[1], this.zoomLevel);
        }
    }
    class Scene extends LocatedMessenger {
        sceneId;
        channels;
        channelkeys;
        label;
        collectionId;
        didWarn;
        static async load(src) {
            const scene = new Scene();
            const payload = await scene.runCommandAsync('open', src);
            scene.afterOpening(payload);
            return scene;
        }
        static fromCollectionFrame(config) {
            const scene = new Scene();
            scene.afterOpening(config);
            return scene;
        }
        afterOpening(payload) {
            this.W = payload['W'];
            this.H = payload['H'];
            this.zoomLevel = payload['zoomLevel'];
            this.worldCoords = payload['worldCoords'];
            this.label = payload["label"] || "";
            this.sceneId = payload['sceneId'];
            this.channels = payload['channels'];
            this.channelkeys = payload['channels'].map(c => `${this.getId()}_${c}`);
            globalChannelkeys = globalChannelkeys.concat(this.channelkeys);
        }
        writePos(channel, pos, val) {
            const channelkey = `${this.getId()}_${channel}`;
            if (this.warnIfNotExists(channelkey))
                return;
            writeShared(channelkey, pos, val);
            //if read first performance is about the same
            return writeShared("dirtymarkers", globalChannelkeys.indexOf(channelkey), 1);
        }
        readPos(channel, pos) {
            const channelkey = `${this.getId()}_${channel}`;
            if (this.warnIfNotExists(channelkey))
                return;
            return readShared(channelkey, pos);
        }
        warnIfNotExists(channelkey) {
            const missingChannel = !globalChannelkeys.includes(channelkey);
            if (!this.didWarn && missingChannel) {
                this.didWarn = true;
                warn(`This scene does not include channel: ${channelkey.split("_")[1]}`);
            }
            return missingChannel;
        }
        async addOverlay(label = "") {
            const overlay = new Overlay();
            const payload = await overlay.runCommandAsync('addOverlay', [this.getId(), label]);
            overlay.afterOpening(payload);
            return overlay;
        }
        forceChannel(channelname) {
            return this.runCommandAsync('forceChannel', [this.getId(), channelname]);
        }
        contains(x, y) {
            return x >= 0 && x < this.W && y >= 0 && y < this.H;
        }
        getId() { return this.sceneId; }
        getWidth() { return this.W; }
        getHeight() { return this.H; }
        getDimensions() { return [this.W, this.H]; }
        getPixel(x, y) {
            return new Pixel(this, x, y);
        }
        getPixels() {
            let x = 0;
            let y = 0;
            let scene = this;
            return {
                [Symbol.iterator]() { return this; },
                next() {
                    if (y < scene.H) {
                        const p = new Pixel(scene, x, y);
                        x = (x + 1) % scene.W;
                        if (x === 0)
                            y++;
                        return { value: p, done: false };
                    }
                    else {
                        return { done: true };
                    }
                }
            };
        }
        get pixels() {
            return this.getPixels();
        }
        getPixelSize() {
            return this.metersPerPixel(this.worldCoords[0] + 0.5 * this.W, this.worldCoords[1] + 0.5 * this.H);
        }
        setChannelAt(channel, x, y, val) {
            const pos = x + y * this.getWidth();
            return this.writePos(channel, pos, val);
        }
        getChannelAt(channel, x, y) {
            const pos = x + y * this.getWidth();
            return this.readPos(channel, pos);
        }
    }
    class Overlay extends Scene {
        setChannelAt(channel, x, y, val) {
            const pos = x + y * this.getWidth();
            if (this.readPos("a", pos) === 0)
                this.writePos("a", pos, 255);
            return this.writePos(channel, pos, val);
        }
        setColorAt(x, y, color) {
            const pos = x + y * this.getWidth();
            if (this.readPos("a", pos) === 0)
                this.writePos("a", pos, 255);
            for (let i = 0; i < color.length; i++) {
                this.writePos(this.channels[i], pos, color[i]);
            }
        }
        addMarker(x, y) {
            const R = Math.round(Math.min(this.W, this.H) / 80);
            for (let dx = -2 * R; dx <= 2 * R; dx++) {
                for (let dy = -2 * R; dy <= 2 * R; dy++) {
                    if (this.contains(x + dx, y + dy)) {
                        const r = (dx ** 2 + dy ** 2) ** 0.5;
                        const dr = Math.abs(R - r);
                        if (dr < 0.3 * R) {
                            this.setChannelAt("r", x + dx, y + dy, 255);
                            this.setChannelAt("a", x + dx, y + dy, dr < 0.2 * R ? 255 : 255 - Math.round(255 * (dr - 0.2 * R) / (0.1 * R)));
                        }
                    }
                }
            }
        }
    }
    class Collection extends LocatedMessenger {
        collectionId;
        scenes;
        afterOpening(payload) {
            this.W = payload['W'];
            this.H = payload['H'];
            this.zoomLevel = payload['zoomLevel'];
            this.worldCoords = payload['worldCoords'];
            this.collectionId = payload['collectionId'];
            let sceneIds = payload["sceneIds"];
            this.scenes = sceneIds.map((sceneId, index) => {
                const scene = Scene.fromCollectionFrame({
                    "W": payload["W"],
                    "H": payload["H"],
                    "worldCoords": payload['worldCoords'],
                    "zoomLevel": payload['zoomLevel'],
                    sceneId,
                    "channels": payload["channels"][index],
                    "label": payload["labels"][index]
                });
                scene.collectionId = this.collectionId;
                return scene;
            });
        }
        static async load(src) {
            const collection = new Collection();
            const payload = await collection.runCommandAsync('openCollection', src);
            collection.afterOpening(payload);
            return collection;
        }
        async addOverlay(label = "") {
            const overlay = new Overlay();
            const payload = await overlay.runCommandAsync('addCollectionOverlay', [this.getId(), label]);
            overlay.afterOpening(payload);
            overlay.collectionId = this.getId();
            return overlay;
        }
        forceFrame(sceneOrLabel) {
            const scene = sceneOrLabel instanceof Scene ? sceneOrLabel : this.scenes.find(s => s.label === sceneOrLabel);
            this.runCommandAsync('forceFrame', [this.getId(), this.scenes.indexOf(scene)]);
        }
        getId() { return this.collectionId; }
        getScenes() { return this.scenes; }
    }
    class Pixel {
        scene;
        x;
        y;
        p;
        constructor(scene, x, y) {
            this.scene = scene;
            this.x = x;
            this.y = y;
            this.p = x + y * scene.getWidth();
        }
        setChannel(channel, value) {
            return this.scene.setChannelAt(channel, this.x, this.y, value);
        }
        getChannel(channel) {
            return this.scene.getChannelAt(channel, this.x, this.y);
        }
        getSizeInMeters() {
            return this.scene.metersPerPixel(this.x, this.y);
        }
        get size() {
            return this.getSizeInMeters();
        }
        getAreaInKm2() {
            return 0.000001 * this.size * this.size;
        }
        get people() {
            return (this.getChannel("pop") || 0) * this.getAreaInKm2();
        }
        getBrightness() {
            return (this.getChannel("r") + this.getChannel("g") + this.getChannel("b")) / 3;
        }
        get brightness() {
            return this.getBrightness();
        }
        toString() {
            const obj = Object.fromEntries(this.scene.channels.map(c => [c, this.scene.getChannelAt(c, this.x, this.y)]));
            return `Pixel(${this.x}, ${this.y}): ${JSON.stringify(obj)}`;
        }
        getNeighbours(includeDiagonal = false) {
            const addPixelIfExists = (list, x, y) => {
                if (this.scene.contains(x, y))
                    list.push(this.scene.getPixel(x, y));
            };
            const neighbours = [];
            let x = this.x;
            let y = this.y;
            addPixelIfExists(neighbours, x, y - 1);
            addPixelIfExists(neighbours, x, y + 1);
            addPixelIfExists(neighbours, x - 1, y);
            addPixelIfExists(neighbours, x + 1, y);
            if (includeDiagonal) {
                addPixelIfExists(neighbours, x - 1, y - 1);
                addPixelIfExists(neighbours, x - 1, y + 1);
                addPixelIfExists(neighbours, x + 1, y - 1);
                addPixelIfExists(neighbours, x + 1, y + 1);
            }
            return neighbours;
        }
        getNeighbors(includeDiagonal = false) {
            return this.getNeighbours(includeDiagonal);
        }
    }
    for (let channelAlias of Object.entries(channelAliases)) {
        const [name, aliases] = channelAlias;
        for (let alias of [...new Set(aliases)]) {
            Object.defineProperty(Pixel.prototype, alias, {
                get() { return this.getChannel(name); },
                set(val) { return this.setChannel(name, val); }
            });
        }
    }
    return { Scene, Pixel, Collection };
}

File: src/skulpthelpers/myplanet.py
from time import sleep, clock
from mercatorWorker import worldPixelToPosition, metersPerPixel

def _asyncWaitFor(signal):
  answer = waitFor(signal)
  while not answer:
    answer = waitFor(signal)
    sleep(0.002)
  return answer['payload']

globalChannelKeys = []

class Messenger:

	def waitForSignal(self, signal):
		return _asyncWaitFor(signal)

	def runCommand(self, command, value):
		signal = 'image'+command+str(round(clock()*1e6))
		show({'command': command, 'value': value, 'readysignal': signal})
		return self.waitForSignal(signal)

	def runCommandImmediate(self, command, value):
		signal = False
		show({'command': command, 'value': value, 'readysignal': signal})


class LocatedMessenger(Messenger):

	def getPosition(self, px, py):
		return worldPixelToPosition(self.worldCoords[0] + px, self.worldCoords[1] + py, self.zoomLevel)

	def metersPerPixel(self, px, py):
		return metersPerPixel(self.getPosition(px, py)[1], self.zoomLevel)

class Scene(LocatedMessenger):

	@classmethod
	def load(cls, src):
		scene = cls()
		payload = scene.runCommand('open', src)
		scene._afterOpening(payload)
		return scene

	@classmethod
	def fromCollectionFrame(cls, config):
		scene = cls()
		scene._afterOpening(config)
		return scene

	def _afterOpening(self, payload):
		global globalChannelKeys
		self.W = payload['W']
		self.H = payload['H']
		self.zoomLevel = payload['zoomLevel']
		self.worldCoords = payload['worldCoords']
		self.label = payload.get("label", "")
		self.sceneId = payload['sceneId']
		self.channels = payload['channels']
		self.channelkeys = [str(self.getId()) + "_" + c for c in self.channels]
		globalChannelKeys += self.channelkeys

	def _writePos(self, channel, pos, val):
		global writeShared
		global globalChannelKeys
		channelkey = str(self.getId()) + "_" + channel
		writeShared(channelkey, pos, val)
		return writeShared("dirtymarkers", globalChannelKeys.index(channelkey), 1)

	def _readPos(self, channel, pos):
		global readShared
		return readShared(str(self.getId()) + "_" + channel, pos)

	def addOverlay(self, label=""):
		overlay = Overlay()
		payload = overlay.runCommand('addOverlay', [self.getId(), label])
		overlay._afterOpening(payload)
		return overlay

	def forceChannel(self, channelname):
		self.runCommand('forceChannel', [self.getId(), channelname])

	def contains(self, x, y):
		return 0 <= x < self.W and 0 <= y < self.H

	def getId(self):
		return self.sceneId

	def getWidth(self):
		return self.W

	def getHeight(self):
		return self.H

	def getDimensions(self):
		return [self.W, self.H]

	def getPixel(self, x, y):
		return Pixel(self, x, y)

	def getPixels(self):
		for y in range(self.H):
			for x in range(self.W):
				yield Pixel(self, x, y)

	@property
	def pixels(self):
		return self.getPixels()

	def getPixelSize(self):
		return self.metersPerPixel(self.worldCoords[0] + 0.5*self.W, self.worldCoords[1] + 0.5*self.H)
	
	def setChannelAt(self, channel, x, y, val):
		if(not channel in self.channels):
			raise Exception("Channel not available: " + channel)
		pos = x + y * self.getWidth()
		return self._writePos(channel, pos, val)

	def getChannelAt(self, channel, x, y):
		if(not channel in self.channels):
			raise Exception("Channel not available: " + channel)
		pos = x + y * self.getWidth()
		return self._readPos(channel, pos)

class Overlay(Scene):
	def setChannelAt(self, channel, x, y, val):
		pos = x + y * self.getWidth()
		if self._readPos("a", pos) == 0:
			self._writePos("a", pos, 255)
		return self._writePos(channel, pos, val)

	def setColorAt(self, x, y, color):
		pos = x + y * self.getWidth()
		if self._readPos("a", pos) == 0:
			self._writePos("a", pos, 255)
		for channel in range(len(color)):
			writeShared(self.channelkeys[channel], pos, color[channel])
		return

	def addMarker(self, x, y):
		R = round(min(self.W, self.H) / 80)
		for dx in range(-2*R, 2*R+1):
			for dy in range(-2*R, 2*R+1):
				if self.contains(x + dx, y + dy):
					r = (dx**2 + dy**2)**0.5
					dr = abs(R-r)
					if dr < 0.3*R:
						self.setChannelAt("r", x + dx, y + dy, 255)
						alpha = 255
						if dr > 0.2 * R:
							alpha -= round(255 * (dr - 0.2*R) / (0.1*R))
						self.setChannelAt("a", x + dx, y + dy, alpha)

class Collection(LocatedMessenger):

	@classmethod
	def load(cls, url):
		collection = cls()
		payload = collection.runCommand('openCollection', url)
		collection._afterOpening(payload)
		return collection

	def _afterOpening(self, payload):
		global globalChannelKeys
		self.W = payload['W']
		self.H = payload['H']
		self.zoomLevel = payload['zoomLevel']
		self.worldCoords = payload['worldCoords']
		self.collectionId = payload['collectionId']
		self.scenes = []
		for index, sceneId in enumerate(payload['sceneIds']):
			scene = Scene.fromCollectionFrame({
				"W": payload["W"],
				"H": payload["H"],
				"worldCoords": payload['worldCoords'],
				"zoomLevel": payload['zoomLevel'],
				"sceneId": sceneId,
				"channels": payload["channels"][index],
				"label": payload["labels"][index]
			})
			scene.collectionId = self.collectionId
			self.scenes.append(scene)

	def addOverlay(self, label = ""):
		overlay = Overlay()
		payload = overlay.runCommand('addCollectionOverlay', [self.getId(), label])
		overlay._afterOpening(payload)
		overlay.collectionId = self.getId()
		return overlay

	def findSceneByLabel(self, label):
		for scene in self.scenes:
			if scene.label == label:
				return scene

	def forceFrame(self, sceneOrLabel):
		scene = sceneOrLabel if isinstance(sceneOrLabel, Scene) else self.findSceneByLabel(sceneOrLabel)
		self.runCommand('forceFrame', [self.getId(), self.scenes.index(scene)])

	def getId(self):
		return self.collectionId

	def getScenes(self):
		return self.scenes


class Pixel:
	def __init__(self, scene, x, y):
		self.scene = scene
		self.__x = x
		self.__y = y
		self.__p = x + y * scene.getWidth()
	
	@classmethod
	def create_channel_methods(cls):
		channelAliases = {
			"r": ["r", "R", "red", "rot"],
			"g": ["g", "G", "green", "gruen", "grün"],
			"b": ["b", "B", "blue", "blau"],
			"a": ["a", "A", "alpha", "alpha"],
			"elevation": ["h", "H", "elevation", "hoehe", "höhe"],
			"country": ["c", "C", "country", "land"],
			"pop": ["p", "P", "pop", "bev"],
			"nox": ["n", "N",  "nox", "NOx"],
			"night": ["l", "L", "night", "nacht", "light", "nightlight", "licht"],
		}
		def createGetter(channel):
			return lambda self: self.getChannel(channel)
		def createSetter(channel):
			return lambda self, val: self.setChannel(channel, val)
		for name, aliases in channelAliases.items():
			for alias in set(aliases):
				setattr(cls, alias, property(createGetter(name), createSetter(name)))
	
	def __getx(self):
		return self.__x
	x = property(__getx)
	
	def __gety(self):
		return self.__y
	y = property(__gety)
	
	def setChannel(self, channel, value):
		return self.scene.setChannelAt(channel, self.x, self.y, value)

	def getChannel(self, channel):
		return self.scene.getChannelAt(channel, self.x, self.y)

	def getSizeInMeters(self):
		return self.scene.metersPerPixel(self.x, self.y)
	size = property(getSizeInMeters)
		
	def getAreaInKm2(self):
		return 0.000001 * self.size * self.size

	def countPeople(self):
		return (self.getChannel("pop") or 0) * self.getAreaInKm2()
	people = property(countPeople)

	def getBrightness(self):
		return (self.getChannel("r") + self.getChannel("g") + self.getChannel("b")) / 3
	brightness = property(getBrightness)

	#alias
	def set(self, channel, value):
		return self.scene.setChannelAt(channel, self.x, self.y, value)

	#alias
	def get(self, channel):
		return self.scene._readPos(channel, self.__p)

	def __str__(self):
		log = "Pixel(%d,%d): {" % (self.x, self.y)
		for channel in self.scene.channels:
			log += channel + ":" + str(self.getChannel(channel)) + "  "
		log += "}"
		return log
	
	def getNeighbours(self):
		def addPixelIfExists(list, x, y):
			if self.scene.contains(x, y):
				list.append(self.scene.getPixel(x, y)) 
		neighbors = []
		x = self.x
		y = self.y
		addPixelIfExists(neighbors, x, y-1)
		addPixelIfExists(neighbors, x, y+1)
		addPixelIfExists(neighbors, x-1, y)
		addPixelIfExists(neighbors, x+1, y)
		return neighbors

	#alias
	def getNeighbors(self):
		return self.getNeighbors()


Pixel.create_channel_methods()